### PR TITLE
Add org.okstar.ok-msg

### PR DESCRIPTION
**Pull request by Joy1024**
### Please confirm your submission meets all the criteria

- [X] Please describe your application briefly.

OkMSG is an enterprise communication collaboration tool developed and maintained by the OkStar (okstar. org) community that focuses on data security and protection. It supports independent private deployment and is an open-source project that integrates multiple functions such as instant messaging, voice, video calling, file sending, and meetings. It enables your enterprise to more effectively initiate collaboration, communicate effectively, control costs, explore new businesses, and help you accelerate business development.

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I have [built][build] and tested the submission locally.
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [ ] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

File: org.okstar.ok-msg.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  - --socket=pulseaudio
+  - --device=all
+  - --talk-name=org.gnome.Mutter.IdleMonitor
+  - --talk-name=org.kde.StatusNotifierWatcher
+  - --talk-name=com.canonical.AppMenu.Registrar
+  - --talk-name=com.canonical.indicator.application
+  - --talk-name=org.ayatana.indicator.application
+  - --filesystem=xdg-run/pipewire-0
+  - --filesystem=xdg-documents
+  - --unset-env=QT_PLUGIN_PATH
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:21]: Why is this needed? Does it ship its own plugins?

File: org.okstar.ok-msg.yml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    version: '23.08'
+    autodownload: true
+    autodelete: false
+
+build-options:
+  cflags: -DNDEBUG
+  cxxflags: -DNDEBUG
+  build-args:
+    - --share=ipc
+    - --share=network
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:36]:
```suggestion
```

File: org.okstar.ok-msg.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  org.freedesktop.Platform.ffmpeg-full:
+    directory: lib/ffmpeg
+    add-ld-path: .
+    version: '23.08'
+    autodownload: true
+    autodelete: false
+
+build-options:
+  cflags: -DNDEBUG
+  cxxflags: -DNDEBUG
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:33]: Why?

File: org.okstar.ok-msg.yml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  - /share/man
+  - '*.prl'
+  - '*.la'
+  - '*.a'
+
+modules:
+  - name: tcl
+    build-options:
+      strip: false
+      no-debuginfo: true
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:57]: Why is debuginfo disabled here and below?

File: org.okstar.ok-msg.yml (hunk 2 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:36]: Because  at building need to access network to download dependencies , otherwise cannot  pass the build.

Comment (bbhtt) [reply org.okstar.ok-msg.yml:36]: Then it won't build, Flathub builds without network access.

File: org.okstar.ok-msg.yml [hunk 5]
```diff
@@ -0,0 +1,108 @@
+id: org.okstar.ok-msg
```
[the remark this answers is not in the document]
Comment (Joy1024) [reply org.okstar.ok-msg.yml:1]: Resovled it, now redirect to my github project page.

File: org.okstar.ok-msg.yml (hunk 2 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:36]: At building cmake need to access network to download dependencies , otherwise cannot  pass the build.

Comment (hfiguiere) [reply org.okstar.ok-msg.yml:36]: Again, this is not possible and it will fail. You need to download the dependencies ahead of time by having them listed in the manifest.

File: org.okstar.ok-msg.yml (hunk 1 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:21]: Resovled it

File: org.okstar.ok-msg.yml (hunk 3 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:33]: Resovled it

File: org.okstar.ok-msg.yml (hunk 4 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:57]: Resovled it

File: .github/workflows/flatpak.yml
Comment (hfiguiere) [inline .github/workflows/flatpak.yml:1]: this should be removed. It not necessary / doesn't belong.

File: org.okstar.ok-msg.yml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+modules:
+  - name: tcl
+    buildsystem: simple
+    build-commands:
+      - cp -r ../tcl /app
+      - cd /app/tcl/unix; ./configure --prefix=/app --exec-prefix=/app && make && make install
+
+    sources:
+      - type: archive
+        url: https://chuanshaninfo.com/download/tcl8.6.14-src.tar.gz
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:57]: you should use the upstream source, not some random mirror.

File: org.okstar.ok-msg.yml [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        sha256: 5880225babf7954c58d4fb0f5cf6279104ce1cd6aa9b71e9a6322540e1c4de66
+
+  - name: tk
+    buildsystem: simple
+    build-commands:
+      - cd unix && ./configure --prefix=/app --exec-prefix=/app --with-tcl=/app/tcl/unix
+      - cd unix; make && make install
+    sources:
+      - type: archive
+        url: https://chuanshaninfo.com/download/tk8.6.14-src.tar.gz
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:67]: you should use the upstream source, not some random mirror.

File: org.okstar.ok-msg.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  - '*.la'
+  - '*.a'
+
+modules:
+  - name: tcl
+    buildsystem: simple
+    build-commands:
+      - cp -r ../tcl /app
+      - cd /app/tcl/unix; ./configure --prefix=/app --exec-prefix=/app && make && make install
+
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:54]: this is not how you build tcl. 

1. /app is NEVER a builddir
2. you can just use `buidsystem: autotools` and `subdir: unix`

File: org.okstar.ok-msg.yml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    sources:
+      - type: archive
+        url: https://chuanshaninfo.com/download/tcl8.6.14-src.tar.gz
+        sha256: 5880225babf7954c58d4fb0f5cf6279104ce1cd6aa9b71e9a6322540e1c4de66
+
+  - name: tk
+    buildsystem: simple
+    build-commands:
+      - cd unix && ./configure --prefix=/app --exec-prefix=/app --with-tcl=/app/tcl/unix
+      - cd unix; make && make install
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:64]: same as for tcl, almost.

File: org.okstar.ok-msg.yml (hunk 6 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:57]: This project release at sourceforge, it seems can not be downloaded by build(due redirect). so I move it to my website.

Comment (hfiguiere) [reply org.okstar.ok-msg.yml:57]: sourceforge works fine here. that's what matters.

File: org.okstar.ok-msg.yml (hunk 8 above)
Comment (Joy1024) [reply org.okstar.ok-msg.yml:54]: 1. Because tk dependence tcl，so at tk building stage that must to use tcl binaries.
2. I try to use autotools, but tk doesn't access to tcl binaries.
3. Would you show me doc that how to use `subdir: unix`.

Comment (hfiguiere) [reply org.okstar.ok-msg.yml:54]:
```
$ man flatpak-manifest
```

Comment (Joy1024) [reply org.okstar.ok-msg.yml]: I use  ` ${FLATPAK_DEST}` instead of /app  can resolve the problem, so this method is correct?

Comment (hfiguiere) [reply org.okstar.ok-msg.yml]: this does change a single thing related to what I said to do. If you don't believe me, github has a search function that will show you how other package do it for better or worse. (hint: Inkscape does it right)

File: org.okstar.ok-msg.yml [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  - --share=network
+  - --socket=wayland
+  - --socket=fallback-x11
+  - --socket=pulseaudio
+  - --device=all
+  - --talk-name=org.gnome.Mutter.IdleMonitor
+  - --talk-name=org.kde.StatusNotifierWatcher
+  - --talk-name=com.canonical.AppMenu.Registrar
+  - --talk-name=com.canonical.indicator.application
+  - --filesystem=home:rw
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:18]: you NEVER need `:rw`.

File: org.okstar.ok-msg.yml [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  org.freedesktop.Platform.ffmpeg-full:
+    directory: lib/ffmpeg
+    add-ld-path: .
+    version: '23.08'
+    autodownload: true
+    autodelete: false
+  org.freedesktop.Platform.GL.default:
+    version: '23.08'
+    autodownload: true
+    autodelete: false
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:30]:
```suggestion
```

File: org.okstar.ok-msg.yml [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+build-options:
+  strip: false
+  no-debuginfo: true
+  build-args:
+    - --share=network
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:36]: this hasn't been addressed. IT IS NOT AN OPTION.

File: org.okstar.ok-msg.yml [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  - '*.prl'
+  - '*.la'
+  - '*.a'
+
+modules:
+  - name: tcl
+    buildsystem: simple
+    build-commands:
+      - cp -r ../tcl ${FLATPAK_DEST}/
+      - cd ${FLATPAK_DEST}/tcl/unix; ./configure --prefix=${FLATPAK_DEST}/ --exec-prefix=/app && make && make install
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:58]: you haven't done what I suggested. It is still the same thing as before.

`${FLATPAK_DEST}` is NOT a build directory.

File: org.okstar.ok-msg.yml (hunk 9 above)
Comment (hfiguiere) [reply org.okstar.ok-msg.yml:68]: not resolved.

File: org.okstar.ok-msg.yml [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+  - '*.la'
+  - '*.a'
+
+modules:
+  - name: tcl
+    buildsystem: simple
+    subdir: unix
+    build-commands:
+      - ./configure --prefix=${FLATPAK_DEST} --exec-prefix=${FLATPAK_DEST}
+      - make && make install
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:52]: I previously said to use `buildsystem: autotools`. still not done.

File: org.okstar.ok-msg.yml [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    sources:
+      - type: archive
+        url: https://sourceforge.net/projects/tcl/files/Tcl/8.6.14/tcl8.6.14-src.tar.gz
+        sha256: 5880225babf7954c58d4fb0f5cf6279104ce1cd6aa9b71e9a6322540e1c4de66
+
+  - name: tk
+    buildsystem: simple
+    subdir: unix
+    build-commands:
+      - ./configure --prefix=${FLATPAK_DEST} --exec-prefix=${FLATPAK_DEST} --with-tcl=${FLATPAK_DEST}/lib
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:62]: I previously said to use `buildsystem: autotools`. still not done.

File: org.okstar.ok-msg.yml [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+  - name: webrtc
+    builddir: false
+    buildsystem: simple
+    build-commands:
+      - cp -r libwebrtc.a ${FLATPAK_DEST}/lib/
+      - cp -r ../webrtc ${FLATPAK_DEST}/include/
+    sources:
+      - type: archive
+        url: https://github.com/crow-misia/libwebrtc-bin/releases/download/121.6167.5.0/libwebrtc-linux-x64.tar.xz
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:93]: how will that work on aarch64?

WebRTC should be built from source instead.

File: org.okstar.ok-msg.yml [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        sha256: d3cd2a45fe790fa1d91703d10550c9129c8bd6f3adf8a257cdc32f350ec69f27
+
+  - name: ok-gloox
+    buildsystem: cmake-ninja
+    sources:
+      - type: git
+        url: https://github.com/okstar-org/ok-gloox.git
+        branch: main
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:101]: NEVER use git branches.
```suggestion
```

File: org.okstar.ok-msg.yml (hunk 14 above)
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:52]:
```suggestion
    buildsystem: autotools
    subdir: unix
```

File: org.okstar.ok-msg.yml [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+      - type: archive
+        url: https://sourceforge.net/projects/tcl/files/Tcl/8.6.14/tcl8.6.14-src.tar.gz
+        sha256: 5880225babf7954c58d4fb0f5cf6279104ce1cd6aa9b71e9a6322540e1c4de66
+
+  - name: tk
+    buildsystem: simple
+    subdir: unix
+    build-commands:
+      - ./configure --prefix=${FLATPAK_DEST} --exec-prefix=${FLATPAK_DEST} --with-tcl=${FLATPAK_DEST}/lib
+      - make && make install
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:63]:
```suggestion
    buildsystem: autotools
    subdir: unix
```

**Comment by hfiguiere** [comment]
> Please describe le application in English.

This is still not done:

image

File: org.okstar.ok-msg.yml (hunk 10 above)
Comment (hfiguiere) [reply org.okstar.ok-msg.yml:18]: this came back

File: org.okstar.ok-msg.yml (hunk 11 above)
Comment (hfiguiere) [reply org.okstar.ok-msg.yml:30]: this came back

File: org.okstar.ok-msg.yml [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    autodownload: true
+    autodelete: false
+  org.freedesktop.Platform.GL.default:
+    version: '23.08'
+    autodownload: true
+    autodelete: false
+
+build-options:
+  strip: false
+  no-debuginfo: true
```
Comment (hfiguiere) [inline org.okstar.ok-msg.yml:34]:
```suggestion
```

there is no reason to do that.

Comment (bbhtt) [reply org.okstar.ok-msg.yml]: why?

**Comment by bbhtt** [comment]
bot, build org.okstar.ok-msg

**Comment by bbhtt** [comment]
`eu-strip: while opening '/srv/buildbot/worker/build-aarch64-6/build/.flatpak-builder/rofiles/rofiles-UA5P2F/files/lib/libtcl8.6.so': Permission denied`

Maybe check if the lib has some weird permissions.

**Comment by bbhtt** [comment]
Add `chmod 755 /app/lib/libtcl*.so` as a post-install command to fix the permissions.

**Comment by bbhtt** [comment]
bot, build org.okstar.ok-msg

File: org.okstar.ok-msg.yml [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    sources:
+      - type: git
+        url: https://github.com/okstar-org/ok-gloox.git
+        commit: f92ff222e1accd1c0dfc0d4498a4a7a8dbffe5d9
+
+  - name: ok-msg-desktop
+    buildsystem: cmake-ninja
+    post-install: |
+      sudo chmod 755 /app/lib/libtcl*.so
+      sudo chmod 755 /app/lib/libtk*.so
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:102]:
```suggestion
```

File: org.okstar.ok-msg.yml [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+
+modules:
+  - name: tcl
+    buildsystem: autotools
+    subdir: unix
+    sources:
+      - type: archive
+        url: https://sourceforge.net/projects/tcl/files/Tcl/8.6.14/tcl8.6.14-src.tar.gz
+        sha256: 5880225babf7954c58d4fb0f5cf6279104ce1cd6aa9b71e9a6322540e1c4de66
+
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:47]:
```suggestion
  - name: tcl
    buildsystem: autotools
    subdir: unix
    sources:
      - type: archive
        url: https://sourceforge.net/projects/tcl/files/Tcl/8.6.14/tcl8.6.14-src.tar.gz
        sha256: 5880225babf7954c58d4fb0f5cf6279104ce1cd6aa9b71e9a6322540e1c4de66
    post-install:
      - chmod +w /app/lib/libtcl*.so
```

File: org.okstar.ok-msg.yml [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+  - name: tk
+    buildsystem: autotools
+    subdir: unix
+    sources:
+      - type: archive
+        url: https://sourceforge.net/projects/tcl/files/Tcl/8.6.14/tk8.6.14-src.tar.gz
+        sha256: 8ffdb720f47a6ca6107eac2dd877e30b0ef7fac14f3a84ebbd0b3612cee41a94
+
```
Comment (bbhtt) [inline org.okstar.ok-msg.yml:55]:
```suggestion
  - name: tk
    buildsystem: autotools
    subdir: unix
    sources:
      - type: archive
        url: https://sourceforge.net/projects/tcl/files/Tcl/8.6.14/tk8.6.14-src.tar.gz
        sha256: 8ffdb720f47a6ca6107eac2dd877e30b0ef7fac14f3a84ebbd0b3612cee41a94
    post-install:
      - chmod +w /app/lib/libtk*.so

```

**Comment by bbhtt** [comment]
bot, build org.okstar.ok-msg

**Comment by bbhtt** [comment]
You have to add a OARS rating to the metainfo file https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#oars-information

**Comment by bbhtt** [comment]
https://github.com/okstar-org/ok-msg-desktop/blob/d4b2754b8f87708c91d35650a05823acb3068897/lib/xdg/ok-msg-desktop.metainfo.xml.in#L24C16-L24C104 screenshot link here is 404

**Comment by bbhtt** [comment]
bot, build org.okstar.ok-msg

**Comment by bbhtt** [comment]
bot, build org.okstar.ok-msg

**Comment by bbhtt** [comment]
Please test the build and check if it is working.

**Comment by Joy1024** [comment]
> Please test the build and check if it is working.

I use `flatpak install --user https://dl.flathub.org/build-repo/112769/org.okstar.ok-msg.flatpakref ` command to successfully installed my app.

When I run the app on my PC(ubuntu22.04) the error message, But at other os(gentoo linux) is ok.
```
gaojie@root-host:~$ flatpak run --verbose org.okstar.ok-msg
F: No installations directory in /etc/flatpak/installations.d. Skipping
F: Opening system flatpak installation at path /var/lib/flatpak
F: Opening user flatpak installation at path /home/gaojie/.local/share/flatpak
F: Opening user flatpak installation at path /home/gaojie/.local/share/flatpak
F: Opening system flatpak installation at path /var/lib/flatpak
error: app/org.okstar.ok-msg/x86_64/master not installed
```

So, How do I to resolve the error?

**Comment by bbhtt** [comment]
The branch should be `test`: `app/org.okstar.ok-msg/x86_64/test`

Can you remove all existing installations and try again? `flatpak remove org.okstar.ok-msg`

**Comment by Joy1024** [comment]
> The branch from the [test build](https://github.com/flathub/flathub/pull/5143#issuecomment-2183715420) should be `test`: `app/org.okstar.ok-msg/x86_64/test`
> 
> Can you remove all existing installations and try again? `flatpak remove org.okstar.ok-msg`

Thank you, the app can ran correctly.

image

**Comment by bbhtt** [comment]
bot, build org.okstar.ok-msg
